### DISCUSSION TITLE
_App rendering twice when having rewrites in config

### DISCUSSION THREAD

**Discussion by ardaerzin** [vercel/next.js#24893] [Help]
Hi y'all 👋  I noticed while looking at the profiler yesterday that all of our pages were rendering twice on initial load. Then saw that _app component was doing the same. Going backwards, I saw that rewrites in config were causing this. I was able to observe the same behavior on a [brand new project](https://github.com/ardaerzin/nextjs-rewrite-test) I created with `npx create-next-app`. The moment I add rewrites, _app starts rendering twice.

so my question is, is this an expected behavior?

**Comment by lpan42** [comment]
I have the same problem. _app renders twice when having rewrite config. Any explanation or is there other solutions?  Thanks

**Comment by ijjk** [comment]
Hi, yes this is expected for prerendered/automatically statically optimized pages as we attempt to parse any additional query values from the rewrites on the client after hydration so that they are available. You can check if the query is fully updated or not for a page by using `useRouter().isReady` from `next/router` ([related docs](https://nextjs.org/docs/api-reference/next/router#router-object))

**Comment by Cottin** [comment]
Just a note: I get the same render twice behaviour without rewrites in config but when having a query string in the URL. 

`/about` makes `_app` render once
`/about?showMessage=true` makes `_app` render twice where first time `useRouter().isReady` is `false` and `useRouter().query` is `{}`.
